frontend/src/metabase/visualizations/components/settings/ChartSettingFieldsPartition.jsx

```diff
@@ -44,148 +31,121 @@ class ChartSettingFieldsPartition extends React.Component {
-            columnFilter={columnFilter}
-            partitionName={name}
-            columns={value[name]}
-            value={value}
-            getColumnSettingValue={this.getColumnSettingValue}
-            onChangeColumnSetting={this.handleChangeColumnSetting}
-            onEditFormatting={this.handleEditFormatting}
-            updateDisplayedValue={this.updateDisplayedValue}
-            commitDisplayedValue={this.commitDisplayedValue}
-          />
-        ))}
-      </div>
-    );
-  }
-}
+  handleDragEnd = ({ source, destination }) => {
+    if (!source || !destination) {
+      return;
+    }
+    const { value, onChange } = this.props;
+    const { droppableId: sourcePartition, index: sourceIndex } = source;
+    const { droppableId: destinationPartition, index: destinationIndex } =
+      destination;
 
-class PartitionInner extends React.Component {
-  render() {
-    const {
-      columns = [],
-      partitionName,
-      columnFilter,
-      onChangeColumnSetting,
-      getColumnSettingValue,
-      onEditFormatting,
-      updateDisplayedValue,
-      commitDisplayedValue,
-      value,
-      connectDropTarget,
-      title,
-      className,
-    } = this.props;
-    return connectDropTarget(
-      <div className={className}>
-        <Label color="medium">{title}</Label>
-        {columns.length === 0 ? (
-          <EmptyPartition
-            columnFilter={columnFilter}
-            value={value}
-            partitionName={partitionName}
-            updateDisplayedValue={updateDisplayedValue}
-          />
-        ) : (
-          columns.map((col, index) => (
-            <Column
-              key={index}
-              partitionName={partitionName}
-              column={col}
-              index={index}
-              columnFilter={columnFilter}
-              value={value}
-              onChangeColumnSetting={onChangeColumnSetting}
-              getColumnSettingValue={getColumnSettingValue}
-              onEditFormatting={onEditFormatting}
-              updateDisplayedValue={updateDisplayedValue}
-              commitDisplayedValue={commitDisplayedValue}
-            />
-          ))
-        )}
-      </div>,
-    );
-  }
-}
+    if (
+      sourcePartition === destinationPartition &&
+      sourceIndex !== destinationIndex
+    ) {
+      const partition = value[sourcePartition];
+      const partitionDup = [...partition];
 
-const Partition = DropTarget(
-  "columns",
-  {
-    // Using a drop target here is a hack to work around another issue.
-    // The version of react-dnd we're on has a bug where endDrag isn't called.
-    // Drop is still called here, so we trigger commit here.
-    drop: (props, monitor, component) => {
-      props.commitDisplayedValue();
-    },
-  },
-  (connect, monitor) => ({ connectDropTarget: connect.dropTarget() }),
-)(PartitionInner);
+      partitionDup[sourceIndex] = partition[destinationIndex];
+      partitionDup[destinationIndex] = partition[sourceIndex];
```

Comment:
It should not be a swap of source and destination because it is not natural:
- Initial list [1, 2, 3]
- Move 1 to the end, expected — [2, 3, 1] 
However, if your swap source and destination the list will become [3, 2, 1]

Reply:
That is true! I'll fix that up